### DISCUSSION TITLE
watchEvents seem to be skipping events from time to time. any ideas?

### DISCUSSION THREAD

**Discussion by ryespresso** [wevm/viem#534] [Question]
For every event, I submit to a webhook. This is in a standard node container. Oddly, the watchEvent log misses some events from time to time.

```
client.watchEvent({
  address: [ADDRESS1, ADDRESS2],
  onError: (error) => console.log(error),
  onLogs: async (logs) => logs.forEach((log) => submit(log)),
});
```

I know this because when I call, the event appears.
```
const logs = await client.getLogs({
    address: [ADDRESS1, ADDRESS2],
    fromBlock: BigInt(0),
  });
```

The client is connected to a goerli fork on anvil. Could this be a bug in watchEvents? Can't really trust the watchEvents function at the moment.

**Comment by jxom** [comment]
The difference between the two is that `watchEvent` will use filters (init w/ `eth_newFilter` then call `eth_getFilterChanges` on a polling interval) and `getLogs` will use `eth_getLogs`. A bit weird because `eth_getFilterChanges` should not skip any logs. Could be an issue in Anvil perhaps.

**Comment by tansanDOTeth** [comment]
@ryeshrimp I have been running watchEvent with WS transport and just storing the logs into redis to verify later and I noticed it misses events. I was using Alchemy at first, but now testing with Infura to see if it might be a JSON RPC issue. 

Did you make any progress on this?

**Comment by tansanDOTeth** [comment]
@jxom Do you know why this might be happening? Same code deployed, but two different web sockets. I would expect it to be consistent between the two.

Alchemy:
<img width="740" alt="CleanShot 2023-08-24 at 11 30 09@2x" src="https://github.com/wagmi-dev/viem/assets/92072158/6efb502f-e51a-41b1-9aba-d65a95c147e6">

Infura: 
<img width="494" alt="CleanShot 2023-08-24 at 11 31 24@2x" src="https://github.com/wagmi-dev/viem/assets/92072158/c504d32b-f58f-4ae5-bb00-472fc2d2e5f9">

**Reply by jxom** [reply]
Could be something weird happening on Infura?

**Reply by tansanDOTeth** [reply]
Not sure. These are infrequent events, so I'll let it run longer and post findings later.

**Reply by tansanDOTeth** [reply]
@jxom Here are some updates and I added http transport for a base case. All 3 are using the same code in a separate server and configured via env vars. Is it possible that the websockets transport connection might be hanging up some how? 

Alchemy Websockets - Last entry:
<img width="706" alt="CleanShot 2023-08-25 at 08 52 05@2x" src="https://github.com/wagmi-dev/viem/assets/92072158/6b64b0a4-911f-488b-9adf-1d5de0ce6a8f">

Alchemy HTTP - After last entry in websockets
<img width="718" alt="CleanShot 2023-08-25 at 08 53 59@2x" src="https://github.com/wagmi-dev/viem/assets/92072158/d385a663-5609-4789-82dd-bd69b8357fd8">

Infura Websockets - Last Entry:
<img width="740" alt="CleanShot 2023-08-25 at 08 55 00@2x" src="https://github.com/wagmi-dev/viem/assets/92072158/b1a4ab2f-bc01-4d2a-9427-5b170d64b7d4">

**Reply by tansanDOTeth** [reply]
Okay, I tested this with the latest version of Viem since the websockets library changed and it still misses events. It almost seems like the websocket connection just silently dies and never reconnects.

It's not very reliable.

**Comment by deanpress** [comment]
> Okay, I tested this with the latest version of Viem since the websockets library changed and it still misses events. It almost seems like the websocket connection just silently dies and never reconnects.
> 
> It's not very reliable.

Can confirm, same issue here. When we use websocket connections, even in fallback mode with both http and websocket endpoints, it appears to fail silently and our service starts missing events.

**Reply by venticello** [reply]
The main problem is that the websocket closes the connection abnormally. Those. it does not report that the connection is closed; it simply stops providing data. Until they add hearbeat here, this will continue to happen. This is a problem with any provider, at least in the polygon network.

When starting the ws transport we need something like this.


```js
/**
   * Starts a heartbeat that pings the websocket server periodically to ensure
   * that the connection stays open.
   *
   * @internal
   */
  private startHeartbeat(): void {
    if (this.heartbeatIntervalId != null) {
      return;
    }
    this.heartbeatIntervalId = setInterval(async () => {
      try {
        await withTimeout(this.send('net_version'), HEARTBEAT_WAIT_TIME);
      } catch {
        this._websocket.reconnect();
      }
    }, HEARTBEAT_INTERVAL);
  }
```

And then make a backfill of logs, you can see the implementation in [Alchemy-SDK](https://github.com/alchemyplatform/alchemy-sdk-js/blob/main/src/api/alchemy-websocket-provider.ts)

Without this, it will not be possible to use this library for full logs.

**Reply by Sa7cez** [reply]
Any news? I use Ankr nodes for listening 10 contracts on 6 chains and have same problem (

**Comment by runjuu** [comment]
We just encountered similar issues, but we are using HTTP instead of WebSocket while watching the events emitted from [Crossbell](https://github.com/wevm/viem/blob/234f8c209fed23a2caa41fbc049909f24a14850f/src/chains/definitions/crossbell.ts#L3).

I found that the reason events are lost is that when the previous filter was invalid, the newly regenerated filter also incurred the  `InvalidInputRpcError`. Then, it [regenerates the filter](https://github.com/wevm/viem/blob/234f8c209fed23a2caa41fbc049909f24a14850f/src/actions/public/watchEvent.ts#L291-L294), resulting in `InvalidInputRpcError` again and again, until at some point, everything returns to normal. I’m not sure why it happens, but during the regeneration period, there is a high chance of losing some events.

To avoid missing events, perhaps the only safe method to obtain and watch them is by using [getLogs](https://viem.sh/docs/actions/public/getLogs.html#getlogs), keeping track of the block number, and repeatedly checking for new logs. This is similar to the fallback logic of `watchEvents`, which occurs when the RPC provider does not support filters. It would be super helpful if there were a way for us to specify which method to use, [by filters](https://github.com/wevm/viem/blob/234f8c209fed23a2caa41fbc049909f24a14850f/src/actions/public/watchEvent.ts#L248-L254) or [by block numbers](https://github.com/wevm/viem/blob/234f8c209fed23a2caa41fbc049909f24a14850f/src/actions/public/watchEvent.ts#L254-L285), when using `watchEvents`.
